src/python/gem5/simulate/multi_sim.py

```diff
@@ -0,0 +1,113 @@
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import itertools
+import os
+import signal
+from functools import partial
+from time import sleep
+from typing import (
+    Any,
+    Callable,
+    List,
+    Tuple,
+)
+
+from gem5.resources.resource import AbstractResource
+from gem5.utils.multiprocessing import Process
+
+from .simulator import Simulator
+
+
+class MultiSim:
+    """
+    This class is used to run multiple simulations in parallel. It takes a list of
+    callables that return a Simulator object and runs them in parallel. The number of
+    simulations that can be run in parallel is limited to the value of the num_cpus parameter
+    """
+
+    def __init__(self, sim_callables: List[Callable[[], Simulator]]):
+        """
+        :param sim_callables: A list of callables that return a Simulator object
+        """
+        self.sim_callables = sim_callables
+
+    def run_all(self, num_cpus: int):
+        """
+        Run all the simulations in parallel
+        :param num_cpus: The number of simulations to run in parallel at a time
+        """
+        processes = []
+
+        for sim_callable in self.sim_callables:
+            if len(processes) > num_cpus:
+                for process in processes:
+                    if not process.is_alive():
+                        processes.remove(process)
+                    os.kill(process.pid, signal.SIGTERM)
+                sleep(1)
+            process_name = (
+                f"{sim_callable.func.__name__}_{sim_callable.args[0].get_id()}"
+            )
+            process = Process(
+                target=run_simulator,
+                args=(sim_callable,),
+                name=process_name,
+            )
+            print(f"Starting process {process_name}")
+
+            process.start()
+            processes.append(process)
+
+        while processes:
+            for process in processes:
+                if not process.is_alive():
+                    processes.remove(process)
+            sleep(1)
+
+        print("All simulations have finished")
+
```

Comment:
You _really_ should've read up on multiprocess `pools` before this. This functionality you provide already exists to handle the scheduling of new processes. The following should be roughly what you need to do:

```py
# Specify the number of processes the pool is allowed to use.
# Note: If `multiprocessing.Pool()` then all available processes are used.
pool = multiprocessing.Pool(processes=num_proc)

# Create a process for each element in `self._sim_callables` passed to function `_run_simulator`. I.e.:
# Process 1 : _run_simulator(self.sim_callables[0])
# Process 2 : _run_simulator(self.sim_callables[1])
# ... and so on.
pool.map(_run_simulator, self.sim_callables) 

# The `pool.map` function will only return when all the processes have finished executing.

def _run_simulator(sim_callable: Callable[[], Simulator]) -> None:
    sim_callable().run()
```

Reply:
Yeah, I had a good reason not to use `multiprocessing.Pool`. It *does* work, but it's going to be limiting our features.

The reason is that I am looking forward to a new feature that wouldn't be possible with `Pool`. I want to be able to track the processes ourselves. Specifically, I want to send a special signal periodically and ask the running simulation to tell me what it's current status is. This is related to the better support for exit events.

I can provide more detail later, but I have a plan!